### DISCUSSION TITLE
How to asign environment variables to the volume's asset filesystem

### DISCUSSION THREAD

**Discussion by goraxan** [craftcms/cms#12478] [Q&A] [answered]
I'm sure this is dumb question but I've been trying to assign an environment variable to a volume's asset filesystem without luck
![image](https://user-images.githubusercontent.com/8337784/208927868-6ae1f434-74df-4626-b854-b8388a5fba21.png)
Whenever I try to put an environment variable, once I focus out the dropdown it gets erased
![image](https://user-images.githubusercontent.com/8337784/208928151-64f26179-ad30-4c2f-9095-9d760814f1dc.png)

Should I set it in some config file? I've read that the config/volumes.php is no longer supported so I don't know how to do it.

Any ideas?

**Comment by brandonkelly** [comment]
The environment variable must already exist, and must be set to the handle of an existing filesystem.

**Reply by goraxan** [reply]
Ok but how can I set the name of the variable in the dropdown? It won't let me introduce it. In my example, MYVARIABLE existed previously.

**Reply by brandonkelly** [reply]
Enter it, prefixed with a `$`. If the environment variable already exists (and was added properly), it will show up in the autosuggest menu as you start typing it.

**Reply by goraxan** [reply]
Well... it's not showing up even though the `FS_AVATARS` variable is correctly set. I'm using Craft CMS 4.3.5 and PHP 8.1.6
This is me trying to set the variable in the dropdown
https://www.loom.com/share/e204e8527fa34c2cb3bcf4add5e3acb0

![image](https://user-images.githubusercontent.com/8337784/209647811-cc61fc61-2b31-44b9-b881-cca0a763d2cb.png)
![image](https://user-images.githubusercontent.com/8337784/209647883-e7036187-409e-49d4-8640-f75a084d2ce1.png)

**Reply by brianjhanson** [reply]
How are you setting the environment variable? We filter out environment values that don't match [over here](https://github.com/craftcms/cms/blob/7ee16681f68f21c1b0e60d2d704c87b5d4e90426/src/web/twig/variables/Cp.php#L588-L608). I'm wondering if there's something we're not accounting for in there.

**Reply by goraxan** [reply]
I see... you are reading from $_SERVER, which looks like it doesn't return the .env file variables (which is where I set the variable). Looks like it only returns PHP system variables
`dump(array_keys($_SERVER))`:
![image](https://user-images.githubusercontent.com/8337784/209790966-e430573b-1df0-488e-a1f0-4cc8dda3c1db.png)
While getenv does return the full list of variables.
`dump(array_keys(getenv()))`:
![image](https://user-images.githubusercontent.com/8337784/209791197-b0148c0d-54af-4444-ac83-46698fc47494.png)

**Reply by goraxan** [reply]
Ok more information... I was serving craft using the `php craft serve` [command ](https://craftcms.com/docs/4.x/console-commands.html#serve).
If I start the PHP built-in server by using the `php -S 0.0.0.0:8080` in the index.php folder, the $_SERVER variable gets populated with the .env variables. That's weird, isn't it?
`dump(array_keys($_SERVER))`:
![image](https://user-images.githubusercontent.com/8337784/209792565-2f0a3ea9-6910-40e6-ba19-aaf466d1628b.png)

**Comment by goraxan** [comment, marked as answer]
Ok... for those who might have the same problem, my problem was caused because of how Dotenv is initialized in the craft starter project I'm using (https://github.com/CreateSean/craft-starter). 
I've solved it by changing both the `craft` and `index.php` files from
`Dotenv\Dotenv::create(CRAFT_BASE_PATH)->load();`
to
`Dotenv\Dotenv::create(CRAFT_BASE_PATH)->overload();`
Related with https://github.com/craftcms/cms/issues/3993#issuecomment-473540306

**Reply by brandonkelly** [reply]
Looks like that starter is still using PHP Dotenv v3: https://github.com/CreateSean/craft-starter/blob/master/composer.json#L32

I’d recommend you update that to v5 (`^5.5.0`), and then change this line in your `craft` executable and `public/index.php` files:

```php
Dotenv\Dotenv::create(CRAFT_BASE_PATH)->load();
```

to:

```php
Dotenv\Dotenv::createUnsafeMutable(CRAFT_BASE_PATH)->safeLoad();
```
